### PR TITLE
Add test for Apple's `-weak_framework` linker argument

### PR DESCRIPTION
**Pull request by madsmtm**
The [`-weak_framework`](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html) linker argument can sometimes be useful to reduce startup times, and to link newer frameworks while still having older deployment targets.

So I made a test to ensure that it continues to work.

Discussed in https://github.com/rust-lang/rust/issues/99427.

### CODE REVIEW

**Comment by madsmtm** [comment]
I guess a test that runs `otool -L` and checks the output could also be useful, are those kinds of tests desired?

**Comment by BlackHoleFox** [comment]
> I guess a test that runs otool -L and checks the output could also be useful, are those kinds of tests desired?

Not your reviewer, but I added tests like that for deployment targets so I don't see why not.

**Comment by madsmtm** [comment]
I've added such tests now, thanks for the input

**Review by b-naber — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I think it failed in https://github.com/rust-lang/rust/pull/119027.

**Comment by madsmtm** [comment]
The `.stderr` test is kinda flaky, since the error output of the linker changes between Xcode versions. I've tried to make it a bit more robust now (tested with Xcode 13.4.1, Xcode 14.3.1 and Xcode 15.0.1), but let me know if you'd rather just have me exclude that part of the test altogether.

**Comment by b-naber** [comment]
Sounds good enough, thanks. 

@bors retry

**Comment by madsmtm** [comment]
Seems like bors didn't pick this up for some reason?

@bors retry

**Comment by madsmtm** [comment]
Ping @b-naber, could you retry bors?

**Comment by fmease** [comment]
#120690
r? compiler

**Review by compiler-errors — APPROVED** [review]

**Comment by madsmtm** [comment]
The `compare-output-lines-by-subset` directive is used rarely enough that it wasn't added to the known directive names, I've added it now.

**Comment by madsmtm** [comment]
The test, as I'd written it before, replaced the string based on the architecture `arm64`, which is why it fails on x86_64. I've made the test generic now, and tested locally with `./x test tests/ui/linkage-attr/framework.rs --target=aarch64-apple-darwin,x86_64-apple-darwin` with 3 different Xcode versions.

File: tests/run-make/link-framework/Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# only-macos
```
Comment (jieyouxu) [inline tests/run-make/link-framework/Makefile:1]: Does this test need to be written in Makefile? Can it be written in `rmake.rs`?

Comment (madsmtm) [reply tests/run-make/link-framework/Makefile:1]: Indeed - didn't know about `rmake.rs`, I think I made this PR before that was available

**Pull request merged**